camera-manager.js

```diff
@@ -121,29 +121,96 @@ const cameraManager = {
       return value1 + (value2 - value1) * amount;
     }
 
+    function checkRayIntersection(origin,dirQuaternion,maxDistance)
+    {
+      let res = false;
+
+      let collision = physicsManager.raycast(origin, dirQuaternion);
+      if (collision) {
+        if (collision.distance <= maxDistance) {
+          res = true;
+        }
+      }
+
+      return res;
+    }
+
+    // Two rays from players position with small offset left and right
+    function checkSmallObjectIntersections(maxDistance)
+    {
+      let intersections = 0;
+
+      const localPlayer = metaversefile.useLocalPlayer();
+
+      let checkPoints = [];
+      
+      const sideways = new THREE.Vector3();
+      
+      // Left
+      checkPoints[0] = localPlayer.position.clone();
+      sideways.set(-1, 0, 0);
+      sideways.applyQuaternion(camera.quaternion);
+      sideways.normalize();
+      checkPoints[0].add(sideways);
+      
+      // Right
+      checkPoints[1] = localPlayer.position.clone();
+      sideways.set(1, 0, 0);
+      sideways.applyQuaternion(camera.quaternion);
+      sideways.normalize();
+      checkPoints[1].add(sideways);
+  
+      for(let i=0;i<checkPoints.length;i++)
+      {
+        rayDirection.subVectors(localPlayer.position,camera.position ).normalize();
+
+        rayMatrix.lookAt(rayDirection,rayVectorZero,rayVectorUp);
+        rayQuaternion.setFromRotationMatrix(rayMatrix);
+  
+        rayStartPos.copy(checkPoints[i]);
+
+        if (checkRayIntersection(rayStartPos,rayQuaternion,maxDistance)) {
+          intersections++;
+        }
+      }
+
+      return intersections;
+    }
+
+
     // Check for collision with level physic objects
 
     let newVal = cameraOffsetTargetZ;
     let hasIntersection = false;
 
     const localPlayer = metaversefile.useLocalPlayer();
 
+    let z = ( camera.near + camera.far ) / ( camera.near - camera.far );
+
     let cornerPoints = [];
     // Top left
-    cornerPoints[0] = new THREE.Vector3( -1, 1, ( camera.near + camera.far ) / ( camera.near - camera.far ) );
+    cornerPoints[0] = new THREE.Vector3( -1, 1, z );
```

Comment:
More allocations.

Reply:
I have removed all allocations from render loop in last commit.